WIP - CASM-4977: Procedure to upgrade NCN and CN with different CSM version based images (#5578)
Cray-HPE/docs-csm: upgrade/Creating_hybrid_configurations.md

# Upgrading management and managed nodes with different CSM versions using IUF

For certain use cases, there may be a need to combine different recipes of CSM to bring up the

> management and managed nodes. For ex: If there is need to use a feature supported for specific

Reviewer: ```suggestion
management and managed nodes. For example, if there is a need to use a feature supported for specific
```

Author: Done

> CPU architecture for managed nodes which is present in a specific recipe of CSM.

Reviewer: ```suggestion
CPU architecture for managed nodes which is present in a different recipe of CSM than the version of CSM that will be run on management nodes.
```

Author: Done

> ## Description

Reviewer: ```suggestion
## Procedure
```

Author: Done

> 1. Decide the recipes for management and managed nodes. For ex: Recipe **`a.b.c`** for CSM version  
> `1.5.x` for managed nodes and recipe **`x.y.z`** for CSM version **`1.6.x`** for management nodes.

Reviewer: ```suggestion
1. Decide the recipes for management and managed nodes. For example, use recipe **`a.b.c`**, which contains CSM version  
`1.5.x`, for managed nodes and use recipe **`x.y.z`**, which contains CSM version **`1.6.x`**, for management nodes.
```

Author: Done

1. Ensure that the CSM cluster is operational with the lowest among the selected recipes for both
management and managed nodes.

> Use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM.

Reviewer: ```suggestion
     If an upgrade to the lowest version of the selected recipes is necessary, use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM and/or products.
```

Author: done

> In the above example, it will be **`a.b.c`** for CSM version **`1.5.x`** for both nodes.

Reviewer: ```suggestion
    In the above example, the upgrade would be to recipe **`a.b.c`**, which contains CSM version **`1.5.x`**, for both management and managed nodes.
```

Author: Done

1. Go through the product versions file for the higher version and come up with a set
of products which have to be upgraded for the recipe. In the above example, the
products from **`x.y.z`** which should be upgraded for CSM version **`1.6.x`**

> should be selected.

Reviewer: Could you reword this a bit? It is unclear to me how you would know which products have to be upgraded.

Author: Please review and confirm if it looks better now

1. Use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM till
[step 9, Managed rollout](../operations/iuf/workflows/managed_rollout.md), skip section

> "2.2 Compute Nodes". That is, do not upgrade compute nodes to the recipe `x.y.z`
> on CSM `1.6.x` and keep the compute nodes on the `a.b.c` recipe on CSM `1.5.x`.`  

Reviewer: ```suggestion
"2.2 Compute Nodes". That is, do not upgrade compute nodes to the recipe `x.y.z`
with CSM version `1.6.x` in order to keep the compute nodes on the `a.b.c` recipe with CSM version `1.5.x`.`  
```

Author: Done

## Post install Testing

> Run required tests to ensure that the upgrade works for the intended use cases.

Reviewer: Could this be more specific? Should the normal post install checks be executed and could a link to the post install checks be added?

Author: Changed

# Upgrading management and managed nodes with different CSM versions using IUF

For certain use cases, there may be a need to combine different recipes of CSM to bring up the

> management and managed nodes. For example, if there is need to use a feature supported for specific
> CPU architecture for managed nodes which is present in a different recipe of CSM than the
> version of CSM that will be run on management nodes.

Reviewer: ```suggestion
management and managed nodes. For example, if a feature which supports a specific
CPU architecture for managed nodes in needed which is present in a different CSM recipe than the CSM recipe that will be run on management nodes.
```

Author: Done

## Procedure

> 1. Decide the recipes for management and managed nodes. For example, use recipe **`a.b.c`**,
> which contains CSM version `1.5.x` for managed nodes and use recipe **`x.y.z`**, which contains
> CSM version **`1.6.x`**, for management nodes.

Reviewer: ```suggestion
1. Decide the recipes for management and managed nodes. For example, use recipe `a.b.c`,
which contains CSM version `1.5.x` for managed nodes and use recipe `x.y.z`, which contains
CSM version `1.6.x`, for management nodes.
```

Author: Done

1. Ensure that the CSM cluster is operational with the lowest among the selected recipes for both
management and managed nodes.

> Use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM.

Reviewer: This should be indented so that it correctly appears under the numbered step above. This will also make it so the numbered steps below don't restart at 1.
```suggestion
If it is necessary to upgrade CSM to the lowest among the selected recipes, then use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM.
```

Author: done

> In the above example, the upgrade would be to recipe **`a.b.c`**, which contains
> CSM version **`1.5.x`**, for both management and managed nodes.

Reviewer: Indent this block as well.
```suggestion
In the above example, the upgrade would be to recipe `a.b.c`, which contains
CSM version `1.5.x`, for both management and managed nodes.
```

Author: Done

1. Go through the `product_vars.yaml` file of both the recipes and come up with a set

> of products and versions across the recipes which have to be upgraded for CSM version **`1.6.x`**.

Reviewer: How do they know which products should be upgraded. Should be all products with a new version in the latest recipe?

> 1. Use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM till

Reviewer: ```suggestion
1. Use the [Upgrade CSM and additional products with IUF](../operations/iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md) procedure to upgrade CSM until
```

Author: Done

[step 9, Managed rollout](../operations/iuf/workflows/managed_rollout.md), skip section
"2.2 Compute Nodes". That is, do not upgrade compute nodes to the recipe `x.y.z`
with CSM version `1.6.x` in order to keep the compute nodes on the `a.b.c` recipe
with CSM version `1.5.x`.

## Post install Testing

> Execute the `ncn healthchecks` to validate that management nodes are executing correctly.

Reviewer: ```suggestion
Execute the NCN health checks to validate that management nodes are working correctly.
```

Author: Done

```bash

>    /opt/cray/tests/install/ncn/automated/ncn-healthcheck-master
>    /opt/cray/tests/install/ncn/automated/ncn-healthcheck-worker
>    /opt/cray/tests/install/ncn/automated/ncn-healthcheck-storage
>    /opt/cray/tests/install/ncn/automated/ncn-kubernetes-checks 

Reviewer: ```suggestion
/opt/cray/tests/install/ncn/automated/ncn-healthcheck-master
/opt/cray/tests/install/ncn/automated/ncn-healthcheck-worker
/opt/cray/tests/install/ncn/automated/ncn-healthcheck-storage
/opt/cray/tests/install/ncn/automated/ncn-kubernetes-checks 
```
Don't need to indent these commands.

Author: Done
